java-checks/src/main/java/org/sonar/java/checks/SpringConstructorInjectionCheck.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S4288")
+public class SpringConstructorInjectionCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return Collections.singletonList(Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if(!hasSemantic()) {
+      return;
+    }
+    ClassTree classTree = (ClassTree) tree;
+    if (isClassTreeAnnotatedWith(classTree,
+      "org.springframework.stereotype.Controller",
+      "org.springframework.stereotype.Repository",
+      "org.springframework.stereotype.Service")) {
+      List<Tree> toReport = classTree.members()
+        .stream()
+        .filter(SpringConstructorInjectionCheck::isMemberAutowired)
+        .map(SpringConstructorInjectionCheck::toReportTree)
+        .collect(Collectors.toList());
+
+      if (!toReport.isEmpty()) {
+        int cost = toReport.size();
+        // find constructor
+        classTree.members().stream()
+          .filter(m -> m.is(Kind.CONSTRUCTOR))
+          .map(m -> ((MethodTree) m).simpleName())
+          .findFirst()
+          .ifPresent(toReport::add);
+
+        reportIssue(toReport.get(0),
```

Comment:
RSPEC is mentioning reporting on annotations. The message is not clear if there is more than one annotation.